channels/channels_test.go

```diff
@@ -168,61 +168,108 @@ func TestChannels(t *testing.T) {
+		err = channelList.DataSent(datatransfer.ChannelID{Initiator: peers[1], Responder: peers[0], ID: tid1}, cids[1], 200, 2, true)
 		require.True(t, xerrors.As(err, new(*channels.ErrNotFound)))
-		require.False(t, isNew)
 
-		isNew, err = channelList.DataReceived(datatransfer.ChannelID{Initiator: peers[0], Responder: peers[1], ID: tid1}, cids[1], 50, 2, true)
+		err = channelList.DataReceived(datatransfer.ChannelID{Initiator: peers[0], Responder: peers[1], ID: tid1}, cids[1], 50, 2, true)
 		require.NoError(t, err)
 		_ = checkEvent(ctx, t, received, datatransfer.DataReceivedProgress)
-		require.True(t, isNew)
 		state = checkEvent(ctx, t, received, datatransfer.DataReceived)
 		require.Equal(t, uint64(100), state.Received())
 		require.Equal(t, uint64(100), state.Sent())
 
-		isNew, err = channelList.DataSent(datatransfer.ChannelID{Initiator: peers[0], Responder: peers[1], ID: tid1}, cids[1], 25, 2, false)
+		err = channelList.DataSent(datatransfer.ChannelID{Initiator: peers[0], Responder: peers[1], ID: tid1}, cids[1], 25, 2, false)
 		require.NoError(t, err)
-		require.False(t, isNew)
 		state = checkEvent(ctx, t, received, datatransfer.DataSent)
 		require.Equal(t, uint64(100), state.Received())
 		require.Equal(t, uint64(100), state.Sent())
 
-		isNew, err = channelList.DataReceived(datatransfer.ChannelID{Initiator: peers[0], Responder: peers[1], ID: tid1}, cids[0], 50, 3, false)
+		err = channelList.DataReceived(datatransfer.ChannelID{Initiator: peers[0], Responder: peers[1], ID: tid1}, cids[0], 50, 3, false)
 		require.NoError(t, err)
-		require.False(t, isNew)
 		state = checkEvent(ctx, t, received, datatransfer.DataReceived)
 		require.Equal(t, uint64(100), state.Received())
 		require.Equal(t, uint64(100), state.Sent())
 	})
 
+	t.Run("data limit", func(t *testing.T) {
+		ds := dss.MutexWrap(datastore.NewMapDatastore())
+
+		channelList, err := channels.New(ds, notifier, decoderByType, decoderByType, &fakeEnv{}, peers[0])
+		require.NoError(t, err)
+		err = channelList.Start(ctx)
+		require.NoError(t, err)
+
+		_, err = channelList.CreateNew(peers[0], tid1, cids[0], selector, fv1, peers[1], peers[0], peers[1])
+		require.NoError(t, err)
+		state := checkEvent(ctx, t, received, datatransfer.Open)
+
+		err = channelList.DataQueued(datatransfer.ChannelID{Initiator: peers[1], Responder: peers[0], ID: tid1}, cids[0], 300, 1, true)
+		require.NoError(t, err)
+		_ = checkEvent(ctx, t, received, datatransfer.DataQueuedProgress)
+		state = checkEvent(ctx, t, received, datatransfer.DataQueued)
+		require.Equal(t, uint64(300), state.Queued())
+
+		err = channelList.SetDataLimit(datatransfer.ChannelID{Initiator: peers[1], Responder: peers[0], ID: tid1}, 400)
+		require.NoError(t, err)
+		state = checkEvent(ctx, t, received, datatransfer.SetDataLimit)
+		require.Equal(t, state.DataLimit(), uint64(400))
+
+		// send block again has no effect
```

Comment:
so .. can you correct me if my interpretation is wrong about what the sequence of initial steps here look like - setting a data limit applies to the entirety of the channel, including data that's already been received, so the `400` limit includes the `300` that's already been received. Then, in this next step, the same data is re-queued but not counted because it's already been counted (aside: why are we able to receive the same data again, and why don't we count duplicate data against the limit? is there abuse possibilities hiding in here?). It's when we get to the next `200` that we go over the `400` and hit a pause. Then the `700` limit set further down applies again to the whole channel, from the start, including the `400` already queued.

Sound about right?

Reply:
yes